modules/protobuf/src/main/scala/smithy4s/protobuf/internals/TaggedCodecSchemaVisitor.scala

```diff
@@ -0,0 +1,285 @@
+            value.toDouble
+          }
+          .addHints(ProtoIndex(2)),
+        Schema.string
+          .oneOf[Document]("string_value", Document.fromString) { case Document.DString(str) => str }
+          .addHints(ProtoIndex(3)),
+        Schema.boolean
+          .oneOf[Document]("bool_value", Document.fromBoolean) { case Document.DBoolean(bool) => bool }
+          .addHints(ProtoIndex(4)),
+        Schema
+          .map(Schema.string, protoJsonSchema)
+          .oneOf[Document]("struct_value", Document.DObject(_)) { case Document.DObject(map) => map }
+          .addHints(ProtoIndex(5), ProtoWrapped()),
+        Schema
+          .indexedSeq(protoJsonSchema)
+          .oneOf[Document]("list_value", Document.DArray(_)) { case Document.DArray(seq) => seq }
+          .addHints(ProtoIndex(6), ProtoWrapped())
+      ) {
+        case DNull       => 0
+        case DNumber(_)  => 1
+        case DString(_)  => 2
+        case DBoolean(_) => 3
+        case DObject(_)  => 4
+        case DArray(_)   => 5
+      }
+      .withId(ShapeId("smithy4s.protobuf", "Json"))
+  }
+
+  def collection[C[_], A](
+      shapeId: ShapeId,
+      hints: Hints,
+      collectionTag: CollectionTag[C],
+      member: Schema[A]
+  ): TaggedCodec[C[A]] = {
+    def drillDownImap[AA, B](taggedCodec: TaggedCodec.IMapCodec[AA, B]): TaggedCodec[C[B]] =
+      taggedCodec.underlying match {
+        case ScalarFieldCodec(scalarCodec) =>
+          PackedRepeatedFieldCodec(scalarCodec.imap(taggedCodec.to, taggedCodec.from), collectionTag)
+        case _ =>
+          UnpackedRepeatedFieldCodec(taggedCodec, collectionTag)
+      }
+
+    val underlying = this(member) match {
+      case ScalarFieldCodec(scalarCodec) => PackedRepeatedFieldCodec(scalarCodec, collectionTag)
+      case imapCodec: IMapCodec[a, A]    => drillDownImap(imapCodec)
+      case other                         => UnpackedRepeatedFieldCodec(other, collectionTag)
+    }
+    if (hints.has(ProtoWrapped)) underlying.wrap else underlying
+  }
+
+  def map[K, V](
+      shapeId: ShapeId,
+      hints: Hints,
+      key: Schema[K],
+      value: Schema[V]
+  ): TaggedCodec[Map[K, V]] = {
+    val underlying = this(
+      Schema
+        .vector(Schema.tuple(key.addHints(ProtoIndex(1)), value.addHints(ProtoIndex(2))))
+        .biject(_.toMap)(_.toVector)
+    )
+    if (hints.has(ProtoWrapped)) underlying.wrap else underlying
+  }
+
+  def enumeration[E](
+      shapeId: ShapeId,
+      hints: Hints,
+      tag: EnumTag[E],
+      values: List[EnumValue[E]],
+      total: E => EnumValue[E]
+  ): TaggedCodec[E] = tag match {
+    case ClosedIntEnum | ClosedStringEnum =>
+      val allIndexed = values.forall(_.hints.has(ProtoIndex))
+      val indexedValues = values.zipWithIndex.map { case (ev, index) =>
+        val protoIndex = if (allIndexed) ev.hints.get(ProtoIndex).map(_.value).getOrElse(index) else index
```

Comment:
**question:** I'm curious about this pattern, it's applied in other kinds of schemas as well.

IIUC the whole conditional is functionally equivalent to the `true` case, so it appears to be an optimization (if everything is indexed we don't have to go through hints at all).
If that's the case, is it really a significant gain, especially given this happens during schema compilation and not on the hot path?

Reply:
It's not really an optimisation, the goal is disregard the manually added indexes if only a subset of the fields are annotated (which should be caught at build time in the nominal case). 

Arguably this falls under the category of "how much effort should we put into validating the schemas at the level of smithy4s ?", and that's a bit of a gray area. Maybe I should throw instead ... ?